WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -125,10 +128,10 @@ public async Task<bool> StartRoundAsync(RoundState roundState, CancellationToken
 			return finalRoundState.WasTransactionBroadcast;
 		}
 
-		private List<AliceClient> CreateAliceClients(RoundState roundState)
+		private ImmutableArray<AliceClient> CreateAliceClients(IEnumerable<Coin> coins, RoundState roundState)
```

Comment:
Why?

Reply:
because that's the narrowest type that makes sense here, and it avoids coupling the API to the implementation details

Reply:
An `ImmutableArray` implements `IEnumerable` what means by changing the method's signature to return `IEnumerable` only restrict what the consumer can do with the result.
Ideally a method should be able to receive very abstract types and return concrete types. In the past we used to return IEnumerables because IReadOnlyCollection was introduced later and was not very popular but now, with immutable collections we have a very good concrete type can can be returned and that guaranties it will not be modified.

Reply:
So it will stay like this - I guess.

Reply:
> Ideally a method should ... return concrete types

why? this appears to be the logic of being co-variant in your arguments and contra-variant in your returns but taken to an extreme.

Being co-variant in the arguments is helpful for a symmetric reason, if you require more abstract types then more concrete types can be substituted in their place when the function is evaluated in a context that has a more abstract type.

But being contra-variant in the return values is not symmetrically helpful, in the same way.

If there is a situation in which the return value of this function is helped by being more concrete, that's a problem with the context the value is being assigned to having too concrete a type.

The purpose of this method is to return a collection of `Alice` objects, and the details of the collection are not relevant. If there is context which receives an `IEnuemrable<Alice>`, this functions' return value will be useful in that context without conversion in both forms, but the `IEnumerable` return will not commit at the type level to implementation details that are not relevant.

Reply:
If I change this I have change this as well: 
`RegisterCoinsAsync(ImmutableArray<AliceClient> aliceClients,`
and 
if (aliceClients.Length == 0)

Reply:
I removed Immutables. It does not make sense and is just the source of delay.